File: server/src/main/java/com/im/server/common/ImPacket.java
package com.im.server.common;

import com.google.protobuf.GeneratedMessageV3;
import io.vertx.core.buffer.Buffer;

import java.nio.ByteBuffer;

public class ImPacket {
    private final int cmd;
    private final int msgType;
    private final int sequenceId;
    private final byte[] bodyBytes;

    public ImPacket(int cmd, int msgType, int sequenceId, byte[] bodyBytes) {
        this.cmd = cmd;
        this.msgType = msgType;
        this.sequenceId = sequenceId;
        this.bodyBytes = bodyBytes;
    }

    public ImPacket(int cmd, int msgType, int sequenceId, GeneratedMessageV3 body) {
        this(cmd, msgType, sequenceId, body.toByteArray());
    }

    public int getCmd() { return cmd; }
    public int getMsgType() { return msgType; }
    public int getSequenceId() { return sequenceId; }
    public byte[] getBodyBytes() { return bodyBytes; }

    public Buffer encode() {
        int totalLen = Cmd.HEADER_LEN + bodyBytes.length;
        ByteBuffer buf = ByteBuffer.allocate(totalLen);
        buf.putShort((short) Cmd.MAGIC);
        buf.putShort((short) Cmd.VERSION);
        buf.putShort((short) cmd);
        buf.put((byte) msgType);
        buf.putInt(sequenceId);
        buf.putInt(bodyBytes.length);
        buf.put(new byte[7]);
        buf.put(bodyBytes);
        return Buffer.buffer(buf.array());
    }
}

File: client-android/app/src/main/java/com/im/client/data/remote/ImPacket.kt
package com.im.client.data.remote

data class ImPacket(
    val cmd: Int,
    val msgType: Int,
    val sequenceId: Int,
    val bodyBytes: ByteArray
) {
    fun encode(): ByteArray {
        val totalLen = Cmd.HEADER_LEN + bodyBytes.size
        val buf = java.nio.ByteBuffer.allocate(totalLen)
        buf.putShort(Cmd.MAGIC.toShort())
        buf.putShort(Cmd.VERSION.toShort())
        buf.putShort(cmd.toShort())
        buf.put(msgType.toByte())
        buf.putInt(sequenceId)
        buf.putInt(bodyBytes.size)
        buf.put(ByteArray(7)) // padding
        buf.put(bodyBytes)
        return buf.array()
    }

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is ImPacket) return false
        return cmd == other.cmd && msgType == other.msgType &&
                sequenceId == other.sequenceId && bodyBytes.contentEquals(other.bodyBytes)
    }

    override fun hashCode(): Int {
        var result = cmd
        result = 31 * result + msgType
        result = 31 * result + sequenceId
        result = 31 * result + bodyBytes.contentHashCode()
        return result
    }
}
